tag/map.go

```diff
@@ -126,13 +174,35 @@ func Update(k Key, v string) Mutator {
 // value of the tag associated with k with v. It inserts the
 // value if k doesn't exist already. It mutates the value
 // if k already exists.
+// Deprecated. Use UpsertWithMetadata instead.
 func Upsert(k Key, v string) Mutator {
 	return &mutator{
 		fn: func(m *Map) (*Map, error) {
 			if !checkValue(v) {
 				return nil, errInvalidValue
 			}
-			m.upsert(k, v)
+			m.upsert(k, v, PropagatingMetadata)
+			return m, nil
+		},
+	}
+}
+
+// UpsertWithMetadata returns a mutator that upserts the
+// value of the tag associated with k with v. It also updates
+// Metadata md. It inserts the value and the metadata if k
+// doesn't exist already. It mutates the value and the
+// Metadata if k already exists.
+// If metadata md is nil then it defaults to NonPropagatingMetadata
+func UpsertWithMetadata(k Key, v string, md *Metadata) Mutator {
```

Comment:
Maybe consider to use the same pattern as StartOption from StartSpan in trace.

```go
type metadatas struct {
	ttl int
}

type Metadata func(*metadatas)

// Not exported for the moment because we want only TTLNoPropagation and TTLUnlimitedPropagation
func withTTL(ttl int) Metadata {
  return func(*metadatas) {
    
  }
}

var (
  TTLNoPropagation = withTTL(valueTTLNoPropagation)
  TTLUnlimitedPropagation = withTTL(valueTTLUnlimitedPropagation)
}

func Upsert(k Key, v string, mds ...Metadata) Mutator {
    metas metadatas
    for _, md := range mds {
        op(&metas)
    }
}
```

Users will use this:
```go
tag.Upsert(key, "my_value", TTLNoPropagation, other_metadata)
```

What do you think? @rakyll do you have any suggestion?